cpp/src/indexing/point_quadtree.cu

```diff
@@ -0,0 +1,957 @@
+               quad_point_count.begin(),
+               quad_point_count.begin() + num_top_quads,
+               lengths->mutable_view().begin<uint32_t>());
+
+  // All leaves are children of the root node (level 0)
+  thrust::fill(policy->on(stream),
+               levels->mutable_view().begin<int8_t>(),
+               levels->mutable_view().end<int8_t>(),
+               0);
+
+  // Quad node indicators are false for leaf nodes
+  thrust::fill(policy->on(stream),
+               indicator->mutable_view().begin<bool>(),
+               indicator->mutable_view().end<bool>(),
+               false);
+
+  // compute f_pos offsets from sizes
+  thrust::exclusive_scan(policy->on(stream),
+                         lengths->view().begin<uint32_t>(),
+                         lengths->view().end<uint32_t>(),
+                         f_pos->mutable_view().begin<uint32_t>());
+
+  std::vector<std::unique_ptr<cudf::column>> cols{};
+  cols.reserve(5);
+  cols.push_back(std::move(keys));
+  cols.push_back(std::move(levels));
+  cols.push_back(std::move(indicator));
+  cols.push_back(std::move(lengths));
+  cols.push_back(std::move(f_pos));
+  return std::make_unique<cudf::experimental::table>(std::move(cols));
+}
+
+template <typename T>
+inline std::unique_ptr<cudf::experimental::table> construct_quadtree(
+  cudf::mutable_column_view &x,
+  cudf::mutable_column_view &y,
+  T const x1,
+  T const y1,
+  T const x2,
+  T const y2,
+  T const scale,
+  cudf::size_type const num_levels,
+  cudf::size_type const min_size,
+  rmm::mr::device_memory_resource *mr,
+  cudaStream_t stream)
+{
+  //
+  // Construct a quad tree from the input (unsorted) x/y points. The bounding
+  // box defined by the x1, y1, x2, and y2 parameters is used to compute keys
+  // in a one-dimensional Z-order curve (i.e. Morton codes) for each point.
+  //
+  // The keys and points are sorted, then the keys are used to construct a
+  // quadtree from the "bottom" level, ascending to the root.
+  //
+  cudf::size_type level_1_size{};
+  cudf::size_type num_top_quads{};
+  cudf::size_type num_parent_nodes{};
+  rmm::device_vector<uint32_t> quad_keys{};
+  rmm::device_vector<uint32_t> quad_point_count{};
+  rmm::device_vector<uint32_t> quad_child_count{};
+  rmm::device_vector<int8_t> quad_levels{};
+
+  //
+  // Wrapped in an IIFE so the temporary `quad_keys`, `quad_point_count`,
+  // `quad_child_count`, and `quad_level` device vectors are freed on return.
+  //
+  // * If `num_parent_nodes` is <= 0, use the temporary vectors to construct a
+  //   leaf-only quadtree.
+  // * If `num_parent_nodes` is >= 1, reverse each level in the temporary
+  //   vectors and return the reversed vectors instead. The reversed vectors are
+  //   used to construct the full quadtree. Freeing the pre-reversed vectors
+  //   early allows RMM to reuse this memory for subsequent temporary
+  //   allocations during full quadtree construction.
+  //
+  std::tie(quad_keys,
```

Comment:
This causes @codereport's "#1 C++ antipattern" (initialize-then-modify), in spite of your use of an IIFE!

Instead, how about unpacking the tuple afterward the call? 

```
auto quads = [&]() {...};

cudf::size_type level_1_size{};
cudf::size_type num_top_quads{std::get<6>(quads)};
cudf::size_type num_parent_nodes{std::get<5>(quads)};
auto quad_keys = std::move(std::get<0>(quads)); // Either use move
auto & quad_child_count = std::get<1>(quads); // or reference variables
// etc.
```